flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/NonTimeRangeUnboundedPrecedingFunction.java

```diff
@@ -0,0 +1,692 @@
+            for (int j = 0; j < ids.size(); j++) {
+                RowData value = valueMapState.get(ids.get(j));
+                collectUpdateBefore(out, value, prevAggValue);
+                collectUpdateAfter(out, value, currAggValue);
+            }
+
+            // update accumulated state for sortKey
+            accMapState.put(curSortKey, currAcc);
+        }
+    }
+
+    private void sendUpdateForChangedRow(
+            Collector<RowData> out,
+            RowKind rowKind,
+            RowData changedRow,
+            RowData prevAggValue,
+            RowData currAggValue) {
+        if (rowKind == RowKind.DELETE) {
+            // TODO: change to collectDelete
+            collectUpdateBefore(out, changedRow, prevAggValue);
+        } else {
+            // For insert, emit either Insert or UpdateAfter
+            collectInsertOrUpdateAfter(out, changedRow, rowKind, currAggValue);
+        }
+    }
+
+    private void collectInsertOrUpdateAfter(
+            Collector<RowData> out, RowData value, RowKind rowKind, RowData currAggValue) {
+        output.setRowKind(rowKind);
+        output.replace(value, currAggValue);
+        out.collect(output);
+    }
+
+    private void collectUpdateBefore(
+            Collector<RowData> out, RowData rowValue, RowData prevAggValue) {
+        output.setRowKind(RowKind.UPDATE_BEFORE);
+        output.replace(rowValue, prevAggValue);
+        out.collect(output);
+    }
+
+    private void collectUpdateAfter(
+            Collector<RowData> out, RowData rowValue, RowData currAggValue) {
+        output.setRowKind(RowKind.UPDATE_AFTER);
+        output.replace(rowValue, currAggValue);
+        out.collect(output);
+    }
+
+    private void collectDelete(Collector<RowData> out, RowData rowValue, RowData prevAggValue) {
+        output.setRowKind(RowKind.DELETE);
+        output.replace(rowValue, prevAggValue);
+        out.collect(output);
+    }
+
+    /**
+     * Removes the row matching delRow from the sortedList. Matching is done based on the sortKeys.
+     * Once matched, the actual row is compared for all ids belonging to the same sortKey. After the
+     * element is removed, a DELETE event is emitted for the removed element, and for all subsequent
+     * elements an UPDATE_BEFORE and UPDATE_AFTER event is emitted based on the previous and newly
+     * aggregated values. Updates are skipped if the previously accumulated value is the same as the
+     * newly accumulated value to save on network bandwidth.
+     *
+     * @param delRow
+     * @param out
+     * @throws Exception
+     */
+    private void removeFromSortedList(RowData delRow, Collector<RowData> out) throws Exception {
+        int i = 0;
+        boolean sortKeyFound = false;
+        delRow.setRowKind(RowKind.INSERT);
+        RowData inputSortKey = sortKeySelector.getKey(delRow);
+        List<Tuple2<RowData, List<Long>>> sortedList = getSortedList();
+
+        for (; i < sortedList.size(); i++) {
```

Comment:
reuse `findIndexForInsertOrUpdate()` here (and give it a more generic name?)

Reply:
Renamed to `findIndexOfSortKey` and reused it in `removedFromSortedList()`